app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.kt

```diff
@@ -0,0 +1,77 @@
+package com.quran.labs.androidquran.widgets
+
+import android.content.Context
+import android.content.Intent
+import android.os.Bundle
+import android.widget.RemoteViews
+import android.widget.RemoteViewsService.RemoteViewsFactory
+import com.quran.data.core.QuranInfo
+import com.quran.labs.androidquran.QuranApplication
+import com.quran.labs.androidquran.R
+import com.quran.labs.androidquran.data.QuranDisplayData
+import com.quran.labs.androidquran.database.BookmarksDBAdapter
+import javax.inject.Inject
+
+class WidgetListProvider(private val context: Context) : RemoteViewsFactory {
+
+  private var suraPageItemList: List<SuraPageItem> = listOf()
+
+  @Inject
+  lateinit var quranInfo: QuranInfo
+
+  @Inject
+  lateinit var quranDisplayData: QuranDisplayData
+
+  init {
+    (context.applicationContext as QuranApplication).applicationComponent.inject(this)
+    populateListItem()
+  }
+
+  private fun populateListItem() {
+    val appContext = context.applicationContext
+    val mBookmarksDBAdapter = BookmarksDBAdapter(appContext, quranInfo.numberOfPages)
+    val bookmarksList = mBookmarksDBAdapter.getBookmarks(BookmarksDBAdapter.SORT_DATE_ADDED)
```

Comment:
I could change this to sort by location if people want, that may seem more sensible.

Reply:
I made this change.